### PR TITLE
fix thread conflicts in getLeaderDist

### PR DESCRIPTION
**Pull request by critical27**
1. Fix thread conflicts when multi-thread emplace in the leader distribution map.
2. Add retry for get leader distribution

### CODE REVIEW

**Comment by critical27** [comment]
Hold this pr for a while, it has another related fix.

**Comment by critical27** [comment]
Ready to go.

**Comment by critical27** [comment]
Jenkins go

File: src/meta/processors/admin/AdminClient.cpp [hunk 1]
```diff
@@ -420,48 +420,67 @@ StatusOr<std::vector<HostAddr>> AdminClient::getPeers(GraphSpaceID spaceId, Part
...
+}
+
 folly::Future<Status> AdminClient::getLeaderDist(HostLeaderMap* result) {
     if (injector_) {
         return injector_->getLeaderDist(result);
     }
     folly::Promise<Status> promise;
     auto future = promise.getFuture();
     auto allHosts = ActiveHostsMan::getActiveHosts(kv_);
+    std::mutex lock;
```
Comment (wadeliuyi) [inline src/meta/processors/admin/AdminClient.cpp]: what about the usage of lock?

File: src/meta/processors/admin/AdminClient.cpp [hunk 2]
```diff
@@ -420,48 +420,67 @@ StatusOr<std::vector<HostAddr>> AdminClient::getPeers(GraphSpaceID spaceId, Part
...
+        for (auto& t : tries) {
+            if (t.hasException()) {
+                continue;
+            }
+            auto&& status = std::move(t.value());
+            if (!status.ok()) {
+                continue;
+            }
+            auto resp = status.value();
+            result->emplace(allHosts[idx++], std::move(resp.leader_parts));
```
Comment (wadeliuyi) [inline src/meta/processors/admin/AdminClient.cpp]: 474行和478行在continue的时候，idx是否也要++，否则，这里取到的leader可能并不是按照idx下标取到的host的leader

File: src/meta/processors/admin/AdminClient.cpp [hunk 3]
```diff
@@ -420,48 +420,67 @@ StatusOr<std::vector<HostAddr>> AdminClient::getPeers(GraphSpaceID spaceId, Part
...
+        folly::Promise<StatusOr<storage::cpp2::GetLeaderResp>> pro;
+        auto fut = pro.getFuture();
+        getLeaderDist(h, std::move(pro), 0, 3);
         hostFutures.emplace_back(std::move(fut));
     }
 
-    folly::collectAll(hostFutures)
-        .then([p = std::move(promise)] (std::vector<folly::Try<Status>>&& tries) mutable {
-        UNUSED(tries);
+    folly::collectAll(hostFutures).then([p = std::move(promise), result, allHosts]
```
Comment (wadeliuyi) [inline src/meta/processors/admin/AdminClient.cpp]: std::move(hostFutures) is better.

File: src/meta/processors/admin/AdminClient.cpp [hunk 4]
```diff
@@ -420,48 +420,67 @@ StatusOr<std::vector<HostAddr>> AdminClient::getPeers(GraphSpaceID spaceId, Part
     return Status::Error("Get Failed");
 }
 
+void AdminClient::getLeaderDist(HostAddr host,
```
Comment (wadeliuyi) [inline src/meta/processors/admin/AdminClient.cpp]: const HostAddr &host is better.

File: src/meta/processors/admin/AdminClient.cpp [hunk 5]
```diff
@@ -420,48 +420,67 @@ StatusOr<std::vector<HostAddr>> AdminClient::getPeers(GraphSpaceID spaceId, Part
...
+                                folly::Promise<StatusOr<storage::cpp2::GetLeaderResp>> pro,
```
Comment (wadeliuyi) [inline src/meta/processors/admin/AdminClient.cpp]: 因为调用的地方是加了std::move了的，一定是一个右值，所以这里修改为folly::Promise<StatusOr<storage::cpp2::GetLeaderResp>> &&pro更好，否则会触发一次右值拷贝构造函数的调用。

File: src/meta/processors/admin/AdminClient.cpp [hunk 6]
```diff
@@ -420,48 +420,67 @@ StatusOr<std::vector<HostAddr>> AdminClient::getPeers(GraphSpaceID spaceId, Part
...
+        storage::cpp2::GetLeaderReq req;
+        auto client = clientsMan_->client(host, evb);
+        client->future_getLeaderPart(std::move(req))
+            .then(evb, [pro = std::move(pro), host, retry, retryLimit, this]
+                       (folly::Try<storage::cpp2::GetLeaderResp>&& t) mutable {
+            if (t.hasException()) {
+                LOG(ERROR) << folly::stringPrintf("RPC failure in AdminClient: %s",
+                                                  t.exception().what().c_str());
+                if (retry < retryLimit) {
+                    usleep(1000 * 50);
```
Comment (wadeliuyi) [inline src/meta/processors/admin/AdminClient.cpp]: maybe add a timer task to delay call getLeaderDist is better than sleep directly.

File: src/meta/processors/admin/AdminClient.cpp (hunk 1 above)
Comment (critical27) [reply src/meta/processors/admin/AdminClient.cpp]: My bad, it should be removed

File: src/meta/processors/admin/AdminClient.cpp (hunk 2 above)
Comment (critical27) [reply src/meta/processors/admin/AdminClient.cpp]: Good catch

File: src/meta/processors/admin/AdminClient.cpp (hunk 3 above)
Comment (critical27) [reply src/meta/processors/admin/AdminClient.cpp]: Good catch.

File: src/meta/processors/admin/AdminClient.cpp (hunk 5 above)
Comment (critical27) [reply src/meta/processors/admin/AdminClient.cpp]: Goot catch.

File: src/meta/processors/admin/AdminClient.cpp (hunk 6 above)
Comment (critical27) [reply src/meta/processors/admin/AdminClient.cpp]: Yeah, but it is not on main data path, we will modify all of them in AdminClient.

File: src/meta/processors/admin/AdminClient.cpp (hunk 4 above)
Comment (critical27) [reply src/meta/processors/admin/AdminClient.cpp]: Good catch

**Pull request merged**